### ISSUE TITLE
borg --stats "All archives" line doesn't really give useful information

### ISSUE THREAD

**Issue by tgharold** [borgbackup/borg#122]
Once you have a few dozen archives in the repository, the "All archives:" line is not very useful.  It starts to get a little silly when you see 20TB original and 10TB compressed and only 100GB deduplicated.  it also fails to give you a feel for "how well are things compressing and deduplicating", which would require measuring the original+deltas rather then naively adding up the raw size all the time.

"Original size" isn't very useful at all.  And I'm not sure how it could be made better.  One idea would be to calculate the original size of a file that has been stored 3x by using it's original size plus the raw size of the two deltas.  So a 100MB file, with a raw delta of 15MB and 10MB would show an original size of 125MB.

"Compressed size" has the same issues.  And should probably calculate off the initial file storage size (compressed) and then add up the sizes of deltas (compressed).  So that 100MB+15MB+10MB file might be 25MB+5MB+2MB when compressed.

Only the "Deduplicated size" column really gives useful information at the moment - as it roughly correlates to the space taken up on the disk by the entire repository.

**Comment by RonnyPfannschmidt** [comment]
Note there are no deltas, reduplication happens at chunk level based on content hashing,

To generate meaningful statistics a full repository scan is needed

**Comment by ThomasWaldmann** [comment]
AFAIK, the values given are correct. Their usefulness might vary, esp. if your expectations about the inner workings of borg differ from how it really works.

Borg does not compute deltas, each backup archive is a full backup (and as far as metadata is concerned, it has ALL the files/directories). But the content data chunks making up the files' contents are reused (and reference counted). But even if you just look at the chunk references: there is no "original" and "reused" chunk reference, all references are created equal.

So "total" is just summing up all the file sizes (like if you did N separate copies of all files). Somehow like having N tar file archives.

"compressed" is similar, just with compression.  Somehow like having N tar.gz file archives.

"deduplicated" is showing how much compressed data chunks are really stored into the repo. Like when using borg.

So, how to improve this?

What we have is a chunk index that has (refcount, size, csize) for each chunk.

Borg currently uses this index to efficiently compute the stats:

total = sum(refcount * size for all chunks)
compressed = sum(refcount * csize for all chunks)
dedup = sum(csize for all chunks)

What could be easily computed additionally is:

dedup_without_compression = sum(size for all chunks)

Obviously this would reflect data compression ratio better than the "compressed" value.

**Comment by tgharold** [comment]
Yeah, I wonder if the following would give more useful information, just dropping the multiplication of the refcount in the loop.

total = sum(size for all chunks)
compressed = sum(csize for all chunks)

Might give a better feel for how much raw chunk data is in the repo, how well it is compressing, and what the final size is.

**Comment by ThomasWaldmann** [comment]
@tgharold yes, but we also need to show the effect of historical/internal/external deduplication.

So I could imagine showing:
total_raw = sum(refcount * size for all chunks)
deduped = sum(size for all chunks)
deduped_compressed = sum(csize for all chunks)

**Comment by oderwat** [comment]
The only stats I would be interested in is how much space a single file/dir/backup uses "total" and what space the deletion of a single archive would free.

**Comment by RonnyPfannschmidt** [comment]
Statistics on chunk reuse might help to understand

**Comment by tgharold** [comment]
That could be an interesting statistic if there was a "borg list --detail" - which would show how much space is taken up by unique-to-that-archive chunks.  That would tell you how much space deletion of that archive would give you.

Alternately, maybe the "All Archives" line should be dropped and replaced by some sort of "Archive Statistics" multi-line section which indicates:

- How many archives there are
- Total uncompressed size = sum(size for all chunks)
- Deduplicated and compressed size = sum(csize for all chunks)

It feels like trying to make the "All Archives" line match the output of the "This archive" line is a bit of a square peg in a round hole.

**Comment by ThomasWaldmann** [comment]
@tgharold "all archives" is just showing off how much progress has been made since the tar days. :)

about "borg list --detail" - yes, that would be nice. and using the cached archive of the single-archive indexes I introduced to make the chunks cache resync more quick, we could do that even rather efficiently.

**Comment by ThomasWaldmann** [comment]
Closing this, considering that master branch is very different from borg 1.x already (and does way less stats).
